Refactor testutil MakePVC to implement builder pattern

AndrewSirenko opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug

/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR refactors test files for external-resizer as @carlory requested in [PR 351](https://github.com/kubernetes-csi/external-resizer/pull/351/files) so that VolumeAttributeClass feature can move to beta.

It makes `MakePVC` simpler by letting developers use use `MakePVC().WithXXX().WithYYY()` to implement an expected PVC.

Finally, this PR refactors helper functions previously in `testutil` to use this pattern and moves them to their respective packages as requested. 

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Partially addresses #367

**Special notes for your reviewer**:
- Will squash commits before merging. 
- Unit tests pass locally. 
- Should I go refactor ALL unit tests to use this pattern? Or is this sufficient for marking this part of #367 off?

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

carlory commented:
/ok-to-test

carlory commented:
@AndrewSirenko Please refactor all the test code with the new library if possible. It makes the test code more readable and maintainable.

AndrewSirenko commented:
Hey @carlory I refactored most of the unit tests and also added in PVWrapper to `testutil` for consistency. I will do a final deep comb through for any places I've missed (like a few lines in `util_test.go`), but would you be able to skim through some of my changes to confirm I'm on the right track? 

Also, I've been keeping the convenience functions like `makeTestPVC` in each package intact. Is this fine or in an ideal world we'd want to get rid of them such that most `WithXXX()`s happen within the testcase definitions?

carlory commented:
Firstly, thanks for working on this task. It's not a blocker for the beta target.

The builder pattern wraps an object inside and returns the inner object after setting the properties via a series of chained methods. The following benefits are:

1. It can return a valid or invalid object. And 🌟🌟🌟  the object may be uncompleted, lacking some required properties but we don't take care of it in our test cases.
2. It is easy to generate the object. so we don't define lots of constructors 🌟🌟 for the object  
3. It provides a helper function to return a wrapper from an existing object. So we can generate a new object with a few properties changed from an existing object.

We can see lots of examples in `k8s.io/kubernetes`. i.e.

- https://github.com/kubernetes/kubernetes/blob/5bf1e95541d90e37f6c6637b5b45d8783e7907aa/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go#L103-L104
- https://github.com/kubernetes/kubernetes/blob/5bf1e95541d90e37f6c6637b5b45d8783e7907aa/pkg/scheduler/framework/plugins/nodename/node_name_test.go#L41

So I added a suggestion to the previous PR and hope to reduce the code duplication in the `pkg/modifycontroller` and `pkg/resizer` directories if possible.

Finally, after I deeply read the code and tried to add some changes, it may increase the complexity and the benefits are not obvious. So we can ignore this suggestion. What do you think? @AndrewSirenko 

cc @msau42

AndrewSirenko commented:
Closing PR, discussed that refactoring to builder pattern may be desirable once all sidecars are in one.
